infra/discovery/src/listener/tools/consistency/ident.js

```diff
@@ -0,0 +1,237 @@
+  )
+  assert(
+    profile.ethAddress.toLowerCase() == dbRecord.ethAddress,
+    `Unexpected ethAddress. ${profile.ethAddress} != ${dbRecord.ethAddress}`
+  )
+
+  // Check attestations
+  for (const attName of ATTESTATION_TYPES) {
+    const attObj = getAttestation(attestations, attName)
+    const attestationValueJSON = get(
+      attObj,
+      `data.attestation.${attName}.verified`
+    )
+    const attestationValueDB = getAttestationDBRepr(dbRecord, attName)
+
+    if (typeof attestationValueJSON === 'undefined') {
+      /**
+       * null from DB, undefined from _.get.  Also, this may be false positive
+       * when things like 'sms' and 'phone' are analogous
+       */
+      if (
+        attestationValueDB !== null &&
+        ['sms', 'phone'].indexOf(attName) < 0
+      ) {
+        assert(
+          attestationValueDB !== null,
+          `DB record value for ${attName} attestation should be null if it isn't included`
+        )
+      }
+    } else {
+      assert(
+        attestationValueJSON === true,
+        `Attestation "verified" should be true or not present`
+      )
+
+      // We can't really verify the values, but we can make sure they exist
+      switch (attName) {
+        case 'facebook':
+        case 'google':
+          assert(
+            attestationValueDB,
+            `DB Value should be true for ${attName} attestation`
+          )
+          break
+        case 'sms':
+        case 'email':
+        case 'phone':
+        case 'twitter':
+        case 'airbnb':
+        default:
+          assert(
+            attestationValueDB !== null,
+            'DB Value should be non-null if attested'
+          )
+      }
+    }
+  }
+}
+
+/**
+ * Verify an identity is what's expected between the IPFS record and database
+ */
+async function verifyIdent(address, ipfsGateway, ipfsHash) {
+  address = address.toLowerCase()
+
+  const records = await db.Identity.findAll({
+    where: {
+      eth_address: address
+    }
+  })
+
+  // Make sure we have the expected amount of records
+  if (records.length < 1) {
+    log.error(`Did not find a matching record for address ${address}`)
+    return false
+  } else if (records.length > 1) {
+    log.warn(`Found too many records for address ${address}`)
+    return false
+  }
+
+  // Fetch the IPFS data
+  let qmHash, ipfsJson
+  try {
+    qmHash = ipfs.getIpfsHashFromBytes32(ipfsHash)
+    ipfsJson = await ipfs.get(ipfsGateway, qmHash)
+  } catch (err) {
+    log.error(`Error retrieving IPFS data for ${qmHash}`)
+    log.error(err)
+  }
+
+  log.debug(`Comparing IPFS Identity ${address} - ${qmHash}`)
+
+  if (!ipfsJson) {
+    log.error(`IPFS Data missing at ${ipfsGateway}/${qmHash}`)
+    return false
+  }
+
+  // Make sure the record is sane
+  try {
+    validateIPFSToDB(ipfsJson, records[0])
+  } catch (err) {
+    // Handle Assertion errors
+    const errString = err.toString()
+    if (errString.indexOf('Assertion') > -1) {
```

Comment:
nit: an alternative would be to define an AssertionError that would be thrown by assert and could be caught here.

Reply:
I was actually looking at this when I was writing it, but it seemed support was iffy.  There's [a lot of non-standard red boxes and considerations in the MDN docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error).  You think that would normally be the way to go in this case?

Reply:
I'm in no way a JS expert :) But my understanding was that custom exceptions are well supported and standard in Node.js. But not so much in browser/babel/etc...

Reply:
I'm going to go ahead and try/do this, since I've never done it in JS before.  Should have it up shortly.

Reply:
Scratch that.  I can't get this to properly work.  Comparing with `instanceof`, same as the MDN example, I just get this:

    [ERROR] event-listener: TypeError: Right-hand side of 'instanceof' is not an object
        at verifyIdent (/home/mike/dev/origin/infra/discovery/src/listener/tools/consistency/ident.js:173:13)
        at processTicksAndRejections (internal/process/task_queues.js:86:5)

And I couldn't find a way to actually get this to work.  If you know something I don't, let me know.

Reply:
Yeah, looking a bit more into that it's actually a bit more convoluted than I thought. But this works:
```
class AssertError extends Error {
  constructor(msg) {
    super(msg)
    this.name = this.constructor.name
   }
 }

try {
  throw new AssertError('boo')
} catch (e) {
  if (e.name === 'AssertError') {
    console.log('AssertError !', e.message)
  }
}
```

Reply:
alright, that'll work